#Exercise 1

In [ ]:
!pip install -q pyomo

In [ ]:
from pyomo.environ import *

In [ ]:
import numpy as np

In [ ]:
import pandas as pd

#Ans 1 :

Decision Variables : $x_{wm}$ : $\text{it denotes the units of truck-loads of WMUs}$

$\text{min} \sum_w\sum_m C[w,m]x_{wm}$

subject to :

$\bullet\sum_w x_{wm}\leq a_w, m = 1,\cdots,k\text{;  }a_w\text{ denotes the available stock at w.}\\
\bullet\sum_m x_{wm}= b_m, w = 1,\cdots,l\text{;  }b_m\text{ denotes the demand at m.}$

In [ ]:
data_csvfile = pd.read_csv('/content/lab5_ex1.csv')

In [ ]:
print(data_csvfile.iloc[0,:])

Warehouses    Ahmedabad
Hassan              627
Jodhpur             617
Trichy             1270
Simla              2072
Vijayawada         1415
Dibrugarh          2943
Raipur             1064
Agartala           4265
Stock               185
Name: 0, dtype: object


In [ ]:
print(data_csvfile.iloc[:,0])

0    Ahmedabad
1    Bengaluru
2      Chennai
3        Delhi
4      Kolkata
5      Lucknow
6       Mumbai
7    Hyderabad
8       Demand
Name: Warehouses, dtype: object


In [ ]:
data_csvfile.columns

Index(['Warehouses', 'Hassan', 'Jodhpur', 'Trichy', 'Simla', 'Vijayawada',
       'Dibrugarh', 'Raipur', 'Agartala', 'Stock'],
      dtype='object')

In [ ]:
data_csvfile.index

RangeIndex(start=0, stop=9, step=1)

In [ ]:
model = ConcreteModel()

In [ ]:
M = len(data_csvfile.index)-1
print('number of warehouses:',M)

number of warehouses: 8


In [ ]:
N = len(data_csvfile.columns)-2
print('number of markets:',N)

number of markets: 8


In [ ]:
row_indices = np.arange(M)
col_indices = np.arange(N)
print(row_indices,col_indices)

[0 1 2 3 4 5 6 7] [0 1 2 3 4 5 6 7]


In [ ]:
for i in row_indices:
  for j in col_indices:
    print('C[',i,'][',j+1,']:',data_csvfile.iloc[i,j+1])

C[ 0 ][ 1 ]: 627
C[ 0 ][ 2 ]: 617
C[ 0 ][ 3 ]: 1270
C[ 0 ][ 4 ]: 2072
C[ 0 ][ 5 ]: 1415
C[ 0 ][ 6 ]: 2943
C[ 0 ][ 7 ]: 1064
C[ 0 ][ 8 ]: 4265
C[ 1 ][ 1 ]: 159
C[ 1 ][ 2 ]: 1432
C[ 1 ][ 3 ]: 351
C[ 1 ][ 4 ]: 3372
C[ 1 ][ 5 ]: 237
C[ 1 ][ 6 ]: 3373
C[ 1 ][ 7 ]: 1456
C[ 1 ][ 8 ]: 5059
C[ 2 ][ 1 ]: 609
C[ 2 ][ 2 ]: 1943
C[ 2 ][ 3 ]: 77
C[ 2 ][ 4 ]: 3365
C[ 2 ][ 5 ]: 298
C[ 2 ][ 6 ]: 3021
C[ 2 ][ 7 ]: 1484
C[ 2 ][ 8 ]: 4803
C[ 3 ][ 1 ]: 2143
C[ 3 ][ 2 ]: 671
C[ 3 ][ 3 ]: 3876
C[ 3 ][ 4 ]: 537
C[ 3 ][ 5 ]: 1902
C[ 3 ][ 6 ]: 1941
C[ 3 ][ 7 ]: 1682
C[ 3 ][ 8 ]: 3348
C[ 4 ][ 1 ]: 2742
C[ 4 ][ 2 ]: 2089
C[ 4 ][ 3 ]: 1931
C[ 4 ][ 4 ]: 2199
C[ 4 ][ 5 ]: 1226
C[ 4 ][ 6 ]: 739
C[ 4 ][ 7 ]: 691
C[ 4 ][ 8 ]: 1128
C[ 5 ][ 1 ]: 2053
C[ 5 ][ 2 ]: 860
C[ 5 ][ 3 ]: 2995
C[ 5 ][ 4 ]: 68
C[ 5 ][ 5 ]: 2069
C[ 5 ][ 6 ]: 2065
C[ 5 ][ 7 ]: 993
C[ 5 ][ 8 ]: 3342
C[ 6 ][ 1 ]: 656
C[ 6 ][ 2 ]: 817
C[ 6 ][ 3 ]: 1245
C[ 6 ][ 4 ]: 2141
C[ 6 ][ 5 ]: 1534
C[ 6 ][ 6 ]: 3351
C[ 6 ][ 7 ]: 1137
C[ 6 ][ 8 ]: 3887
C[ 7 ][ 1 ]

In [ ]:
model.x = Var(row_indices,col_indices,domain=NonNegativeIntegers)

In [ ]:
model.objective = Objective(expr = sum(data_csvfile.iloc[i,j+1]*model.x[i,j] for i in row_indices for j in col_indices),sense = minimize)

In [ ]:
model.constraints = ConstraintList()

In [ ]:
for i in row_indices:
  model.constraints.add(expr=sum(model.x[i,j] for j in col_indices) <= data_csvfile.iloc[i,N+1])

In [ ]:
for j in col_indices:
  model.constraints.add(expr=sum(model.x[i,j] for i in row_indices) == data_csvfile.iloc[M,j+1])

In [ ]:
model.pprint()

4 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   16 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain              : Size : Members
        None :     2 : x_index_0*x_index_1 :   64 : {(0, 0), (0, 1), (0, 2), (0, 3), (0, 4), (0, 5), (0, 6), (0, 7), (1, 0), (1, 1), (1, 2), (1, 3), (1, 4), (1, 5), (1, 6), (1, 7), (2, 0), (2, 1), (2, 2), (2, 3), (2, 4), (2, 5), (2, 6), (2, 7), (3, 0), (3, 1), (3, 2), (3, 3), (3, 4), (3, 5), (3, 6), (3, 7), (4, 0), (4, 1), (4, 2), (4, 3), (4, 4), (4, 5), (4, 6), (4, 7), (5, 0), (5, 1), (5, 2), (5, 3), (5, 4), (5, 5), (5, 6), (5, 7), (6, 0), (6, 1), (6, 2), (6, 3), (6, 4), (6, 5), (6, 6), (6, 7), (7, 0), (7, 1), (7, 2), (7, 3), (7, 4), (7, 5), (7, 6), (7, 7)}
    x_index_0 : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
       

In [ ]:
!apt-get install -y -qq coinor-cbc

In [ ]:
opt_cbc = SolverFactory('cbc')

In [ ]:
result = opt_cbc.solve(model)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [ ]:
print('Objective = ', model.objective())

print('\nDecision Variables')
for i in row_indices:
  for j in col_indices:
    print('Truck-Loads of WMUs transported between ',data_csvfile['Warehouses'][i],' and ', data_csvfile.columns[j+1],':', model.x[i,j].value)

print('\nConstraints')
model.constraints.display()

Objective =  1166465.0

Decision Variables
Truck-Loads of WMUs transported between  Ahmedabad  and  Hassan : 0.0
Truck-Loads of WMUs transported between  Ahmedabad  and  Jodhpur : 185.0
Truck-Loads of WMUs transported between  Ahmedabad  and  Trichy : 0.0
Truck-Loads of WMUs transported between  Ahmedabad  and  Simla : 0.0
Truck-Loads of WMUs transported between  Ahmedabad  and  Vijayawada : 0.0
Truck-Loads of WMUs transported between  Ahmedabad  and  Dibrugarh : 0.0
Truck-Loads of WMUs transported between  Ahmedabad  and  Raipur : 0.0
Truck-Loads of WMUs transported between  Ahmedabad  and  Agartala : 0.0
Truck-Loads of WMUs transported between  Bengaluru  and  Hassan : 0.0
Truck-Loads of WMUs transported between  Bengaluru  and  Jodhpur : 0.0
Truck-Loads of WMUs transported between  Bengaluru  and  Trichy : 0.0
Truck-Loads of WMUs transported between  Bengaluru  and  Simla : 0.0
Truck-Loads of WMUs transported between  Bengaluru  and  Vijayawada : 350.0
Truck-Loads of WMUs transporte

#Ans 7 :

The total cost is : 1166465

From Ahmedabad, 185 truck loads of WMUs are transported to Jodhpur.

From Bengaluru, 350 truck loads of WMUs are transported to Vijaywada.

From Chennai, 375 truck loads of WMUs are transported to Trichy, 30 truck loads of WMUs are transported to Trichy.

From Delhi, 200 truck loads of WMUs are transported to Jodhpur, 80 truck loads of WMUs are transported to Dibrugarh.

From Kolkata, 20 truck loads of WMUs are transported to Dibrugarh, 85 truck loads of WMUs are transported to Agartala.

From Lucknow, 70 truck loads of WMUs are transported to Simla, 5 truck loads of WMUs are transported to Raipur.

From Mumbai, 175 truck loads of WMUs are transported to Hassan, 15 to Jodhpur, 175 to Trichy, and 60 to Raipur.

From Hyderabad, 270 truck loads of WMUs are transported to Vijaywada.

#Ans 8 :

To solve the problem without changing our model, we'd access the corresponding cost of transporting from Bengaluru to Hassan and make it absurdly large, since it is a cost minimization problem, we'll achieve the desired result.

In [ ]:
data_csvfile['Hassan'][1] = 5000
model.objective = Objective(expr = sum(data_csvfile.iloc[i,j+1]*model.x[i,j] for i in row_indices for j in col_indices),sense = minimize)

    'pyomo.core.base.objective.ScalarObjective'>) on block unknown with a new
    Component (type=<class 'pyomo.core.base.objective.ScalarObjective'>). This
    is usually indicative of a modelling error. To avoid this warning, use
    block.del_component() and block.add_component().


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [ ]:
result = opt_cbc.solve(model)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [ ]:
print('Objective = ', model.objective())

print('\nDecision Variables')
for i in row_indices:
  for j in col_indices:
    print('Truck-Loads of WMUs transported between ',data_csvfile['Warehouses'][i],' and ', data_csvfile.columns[j+1],':', model.x[i,j].value)

print('\nConstraints')
model.constraints.display()

Objective =  1166465.0

Decision Variables
Truck-Loads of WMUs transported between  Ahmedabad  and  Hassan : 0.0
Truck-Loads of WMUs transported between  Ahmedabad  and  Jodhpur : 185.0
Truck-Loads of WMUs transported between  Ahmedabad  and  Trichy : 0.0
Truck-Loads of WMUs transported between  Ahmedabad  and  Simla : 0.0
Truck-Loads of WMUs transported between  Ahmedabad  and  Vijayawada : 0.0
Truck-Loads of WMUs transported between  Ahmedabad  and  Dibrugarh : 0.0
Truck-Loads of WMUs transported between  Ahmedabad  and  Raipur : 0.0
Truck-Loads of WMUs transported between  Ahmedabad  and  Agartala : 0.0
Truck-Loads of WMUs transported between  Bengaluru  and  Hassan : 0.0
Truck-Loads of WMUs transported between  Bengaluru  and  Jodhpur : 0.0
Truck-Loads of WMUs transported between  Bengaluru  and  Trichy : 0.0
Truck-Loads of WMUs transported between  Bengaluru  and  Simla : 0.0
Truck-Loads of WMUs transported between  Bengaluru  and  Vijayawada : 350.0
Truck-Loads of WMUs transporte

#Ans 9 :

The total cost is : 1166465

From Ahmedabad, 185 truck loads of WMUs are transported to Jodhpur.

From Bengaluru, 350 truck loads of WMUs are transported to Vijaywada.

From Chennai, 375 truck loads of WMUs are transported to Trichy, 30 truck loads of WMUs are transported to Trichy.

From Delhi, 200 truck loads of WMUs are transported to Jodhpur, 80 truck loads of WMUs are transported to Dibrugarh.

From Kolkata, 20 truck loads of WMUs are transported to Dibrugarh, 85 truck loads of WMUs are transported to Agartala.

From Lucknow, 70 truck loads of WMUs are transported to Simla, 5 truck loads of WMUs are transported to Raipur.

From Mumbai, 175 truck loads of WMUs are transported to Hassan, 15 to Jodhpur, 175 to Trichy, and 60 to Raipur.

From Hyderabad, 270 truck loads of WMUs are transported to Vijaywada.

***Since there was no transportation between Bengaluru and Hassan initially, we do not see any changes in the solution.***

#Ans 10 :

Similarly, we increase the corresponding cost of transporting from Lucknow to Simla and from Kolkata to Agartala absurdly and achieve the desired result.

In [ ]:
data_csvfile['Simla'][5] = 5000
data_csvfile['Agartala'][4] = 5000
model.objective = Objective(expr = sum(data_csvfile.iloc[i,j+1]*model.x[i,j] for i in row_indices for j in col_indices),sense = minimize)

    'pyomo.core.base.objective.ScalarObjective'>) on block unknown with a new
    Component (type=<class 'pyomo.core.base.objective.ScalarObjective'>). This
    is usually indicative of a modelling error. To avoid this warning, use
    block.del_component() and block.add_component().


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [ ]:
result = opt_cbc.solve(model)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [ ]:
print('Objective = ', model.objective())

print('\nDecision Variables')
for i in row_indices:
  for j in col_indices:
    print('Truck-Loads of WMUs transported between ',data_csvfile['Warehouses'][i],' and ', data_csvfile.columns[j+1],':', model.x[i,j].value)

print('\nConstraints')
model.constraints.display()

Objective =  1289875.0

Decision Variables
Truck-Loads of WMUs transported between  Ahmedabad  and  Hassan : 0.0
Truck-Loads of WMUs transported between  Ahmedabad  and  Jodhpur : 185.0
Truck-Loads of WMUs transported between  Ahmedabad  and  Trichy : 0.0
Truck-Loads of WMUs transported between  Ahmedabad  and  Simla : 0.0
Truck-Loads of WMUs transported between  Ahmedabad  and  Vijayawada : 0.0
Truck-Loads of WMUs transported between  Ahmedabad  and  Dibrugarh : 0.0
Truck-Loads of WMUs transported between  Ahmedabad  and  Raipur : 0.0
Truck-Loads of WMUs transported between  Ahmedabad  and  Agartala : 0.0
Truck-Loads of WMUs transported between  Bengaluru  and  Hassan : 0.0
Truck-Loads of WMUs transported between  Bengaluru  and  Jodhpur : 0.0
Truck-Loads of WMUs transported between  Bengaluru  and  Trichy : 0.0
Truck-Loads of WMUs transported between  Bengaluru  and  Simla : 0.0
Truck-Loads of WMUs transported between  Bengaluru  and  Vijayawada : 350.0
Truck-Loads of WMUs transporte

#Ans 11 :

The total cost is : 1289875

From Ahmedabad, 185 truck loads of WMUs are transported to Jodhpur.

From Bengaluru, 350 truck loads of WMUs are transported to Vijaywada.

From Chennai, 375 truck loads of WMUs are transported to Trichy, 30 truck loads of WMUs are transported to Trichy.

From Delhi, 200 truck loads of WMUs are transported to Jodhpur, 70 truck loads of WMUs are transported to Shimla and 10 to Agartala.

From Kolkata, 100 truck loads of WMUs are transported to Dibrugarh, 5 truck loads of WMUs are transported to Raipur.

From Lucknow, 75 trucks loads of WMUs are transported to Agartala.

From Mumbai, 175 truck loads of WMUs are transported to Hassan, 15 to Jodhpur, 175 to Trichy, and 60 to Raipur.

From Hyderabad, 270 truck loads of WMUs are transported to Vijaywada.